packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,273 @@
+                        if (!reusedPasswordItemIds.has(item.id)) {
+                            reusedPasswords.push({ item, vault });
+                            reusedPasswordItemIds.add(item.id);
+                        }
+
+                        auditResults.push({
+                            type: AuditResultType.ReusedPassword,
+                            fieldIndex: passwordField.fieldIndex,
+                        });
+
+                        // Also tag the first matching item as reused, once
+                        const firstMatch = firstMatchPerPasswordHash.get(passwordHash)!;
+                        if (!reusedPasswordItemIds.has(firstMatch.item.id)) {
+                            reusedPasswords.push({ item: firstMatch.item, vault: firstMatch.vault });
+                            reusedPasswordItemIds.add(firstMatch.item.id);
+
+                            auditResults.push({
+                                type: AuditResultType.ReusedPassword,
+                                fieldIndex: firstMatch.passwordField.fieldIndex,
+                            });
+                        }
+                    }
+
+                    firstMatchPerPasswordHash.set(passwordHash, { item, vault, passwordField });
+
+                    // Perform weak audit
+                    const isThisPasswordWeak = await isPasswordWeak(passwordField.field.value);
+                    if (isThisPasswordWeak) {
+                        // Don't add the same item twice to the list, if there are more than one weak password fields in it
+                        if (!weakPasswordItemIds.has(item.id)) {
+                            weakPasswords.push({ item, vault });
+                            weakPasswordItemIds.add(item.id);
+                        }
+
+                        auditResults.push({ type: AuditResultType.WeakPassword, fieldIndex: passwordField.fieldIndex });
+                    }
+
+                    // Perform compromised audit
+                    const isPasswordCompromised = await hasPasswordBeenCompromised(passwordHash);
+                    if (isPasswordCompromised) {
+                        // Don't add the same item twice to the list, if there are more than one compromised password fields in it
+                        if (!compromisedPasswordItemIds.has(item.id)) {
+                            compromisedPasswords.push({ item, vault });
+                            compromisedPasswordItemIds.add(item.id);
+                        }
+
+                        auditResults.push({
+                            type: AuditResultType.CompromisedPassword,
+                            fieldIndex: passwordField.fieldIndex,
+                        });
+                    }
+                }
+
+                item.auditResults = auditResults;
+                item.lastAudited = new Date();
+                vault.items.update(item);
+            }
+        }
+
+        app.save();
```

Comment:
@MaKleSoft this updates the items, but after a sync it's reset, so it doesn't persist. What am I missing?

Reply:
Ah, I forgot about `vault.commit`. Actually, better use `app.saveVault(vault)` (further up, in the outer loop).